rules/import-path-order.js

```diff
@@ -0,0 +1,236 @@
+			return {
+				messageId: MESSAGE_ID_GROUP,
+				data: {
+					earlier: GROUP_NAMES[next.group],
+					later: GROUP_NAMES[prev.group].toLowerCase()
+				}
+			};
+		}
+
+		return null;
+	}
+
+	if (prev.depth !== next.depth) {
+		if (prev.depth < next.depth) {
+			return {
+				messageId: MESSAGE_ID_DEPTH
+			};
+		}
+
+		return null;
+	}
+
+	// TODO: Case insensitive
+	if (next.name < prev.name) {
+		return {
+			messageId: MESSAGE_ID_ORDER
+		};
+	}
+
+	return null;
+}
+
+const create = context => {
+	const {options} = context;
+	const {
+		allowBlankLines = false
+	} = options[0] || {};
+
+	let prevOrder = null;
+	let prevNode = null;
+
+	return {
+		'Program > VariableDeclaration[declarations.length=1] > VariableDeclarator:matches([id.type="Identifier"],[id.type="ObjectPattern"]) > CallExpression[callee.name="require"][arguments.length=1][arguments.0.type="Literal"]': node => {
+			const nextOrder = getOrder(node.arguments[0].value);
+
+			const message = getInvalidOrderMessage(prevOrder, nextOrder);
+
+			if (message) {
+				context.report({
+					node,
+					...message
+				});
+			}
+
+			if (!allowBlankLines) {
+				const blankLinesMessage = getInvalidBlankLinesMessage(prevNode, node, context);
+				if (blankLinesMessage) {
+					context.report({
+						node,
+						...blankLinesMessage
+					});
+				}
+			}
+
+			prevOrder = nextOrder;
+			prevNode = node;
+		},
+		'Program > ExpressionStatement > CallExpression[callee.name="require"][arguments.length=1][arguments.0.type="Literal"]': node => {
+			const nextOrder = getOrder(node.arguments[0].value);
+
+			const message = getInvalidOrderMessage(prevOrder, nextOrder);
+
+			if (message) {
+				context.report({
+					node,
+					...message
+				});
+			}
+
+			if (!allowBlankLines) {
+				const blankLinesMessage = getInvalidBlankLinesMessage(prevNode, node, context);
+				if (blankLinesMessage) {
+					context.report({
+						node,
+						...blankLinesMessage
+					});
+				}
+			}
+
+			prevOrder = nextOrder;
+			prevNode = node;
+		},
+		'Program > ImportDeclaration': node => {
+			const next = getOrder(node.source.value);
+
+			const message = getInvalidOrderMessage(prevOrder, next);
+
+			if (message) {
+				context.report({
+					node: node.source,
+					...message
+				});
+			}
+
+			if (!allowBlankLines) {
+				const blankLinesMessage = getInvalidBlankLinesMessage(prevNode, node, context);
+				if (blankLinesMessage) {
+					context.report({
+						node: node.source,
+						...blankLinesMessage
+					});
+				}
+			}
+
+			prevOrder = next;
+			prevNode = node;
+		}
+	};
+};
+
+module.exports = {
+	create,
+	meta: {
+		type: 'suggestion',
+		docs: {
+			url: getDocsUrl(__filename)
+		},
+		messages: {
+			[MESSAGE_ID_BLANKLINES]: 'Group imports together',
+			[MESSAGE_ID_DEPTH]: 'Relative paths should be sorted by depth',
+			[MESSAGE_ID_GROUP]: '{{earlier}} imports should come before {{later}} imports',
+			[MESSAGE_ID_ORDER]: 'Imports should be sorted alphabetically'
```

Comment:
We didn't discuss this in #324, or I must have missed it. Sorting alphabetically sounds nice in theory, but it creates a lot of annoyance in practice for very little actual benefit.

Reply:
The primary reason I like it is because it puts imports from the same package next to each other. I don't actually care much about the alphabetization; it's just the fastest proxy for doing this:

```
import flat from 'lodash/flat';
import map from 'lodash/map';
import React from 'react';
```

Instead of this:

```
import flat from 'lodash/flat';
import React from 'react';
import map from 'lodash/map';
```

The latter drives me bonkers and I do care pretty strongly about it.

Reply:
Why not just `import {flat, map} from 'lodash';`? I guess the answer is tree-shaking, but I think that also works for named imports.

Even if not, i think it would be better to just place same-named imports together without forcing full alphabetization.

Reply:
In the case of `lodash`, yeah, tree-shaking is the answer. `lodash` is particularly notorious for [not tree-shaking](https://github.com/lodash/lodash/issues?q=is%3Aissue+tree+shaking+is%3Aclosed).

In addition, some libraries split up their common imports into separate packages:

```
import { withRouter } from "react-router";
import { BrowserRouter as Router, Route, Link } from "react-router-dom";
```

----

```
import { makeStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import { AccessAlarm, ThreeDRotation } from '@material-ui/icons';
```

---

I see no real disadvantages to keeping imports alphabetized other than it's mildly annoying. Once the fixer is in place, I don't think it will trip people up too much.

If it really bothers you, I can try to add something that keeps imports clumped by common path part (e.g., `@material-ui` and `core`.) It won't work for things like `react-router` and `react-router-dom` but I can live with that.

Reply:
> I can try to add something that keeps imports clumped by common path part (e.g., @material-ui and core.) It won't work for things like react-router and react-router-dom but I can live with that.

Group modules that have the same super-string? So if you have `react`, then `react-router` and `react-dom` will be grouped with it. That would be generally useful for many other things too, like `electron` projects; `electron`, `electron-util`, `electron-context-menu`, etc.

Reply:
I see four plausible choices. In descending order of complexity:

1. Alphabetize
2. Group identical path segments next to each other (e.g., `@material-ui/core` and `@material-ui/icons`)
3. Group identical path segments _and_ group initial package "word" (e.g., `electron` and `electron-util`)
4. Group identical path segments _and_ group package words (e.g., `foo-bar-baz` would need to be near `foo-bar-zaz`)

(4) smells overly complex and something that will have weird edge cases.

(3) seems fine other than "initial word" is arbitrary. Should `url-join` be grouped next to `url-parse` even though they are technically unrelated? If the description of the rule was (1) than I don't think anyone would be surprised but the intent behind (3) is that libraries are grouped.

(2) is a reasonable compromise.

(1) is my preference because it is simple and covers nearly all of the benefits from the other three options. I don't think anyone would be surprised by it. We can always make it a configuration option, too. (I was planning on adding a case-insensitive switch anyway so we can just make it an `"alphabetize"` key with `"case-sensitive"`, `"case-insensitive"` and `"off"` values.)

Reply:
> Should url-join be grouped next to url-parse even though they are technically unrelated?

I don't think that will happen a lot and it kinda makes sense to though, as both deal with URL. It will however catch a lot of valid cases.

> (4) smells overly complex and something that will have weird edge cases.

Agreed.

---

The problem with 2 is that it doesn't catch common cases. It only catches scoped packages, but the majority of packages today are still top-level packages with a prefix, like `react` and `react-dom`.

So personally, I would go with 2 and 3 on by default, and ability to opt into 1 for people that want it very strict.

Reply:
> So personally, I would go with 2 and 3 on by default, and ability to opt into 1 for people that want it very strict.

Alright; I'll see what I can do.